Задания по 2му спринту

yakovlevyaroslav opened the pull request.

yakovlevyaroslav committed:
Задания по 2му спринту

gennady-bars commented on src/utils/api.js:
```diff
@@ -0,0 +1,23 @@
+export const URL_INGREDIENTS = 'https://norma.nomoreparties.space/api/ingredients';
+export const URL_ORDER = 'https://norma.nomoreparties.space/api/orders';
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars commented on src/utils/api.js:
```diff
@@ -0,0 +1,23 @@
+      throw new Error('Network response was not ok ' + response.statusText);
+    }
+
+    const responseData = await response.json();
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужного повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/utils/api.js:
```diff
@@ -0,0 +1,23 @@
+
+    const responseData = await response.json();
+    return responseData;
+  } catch (error) {
```
Не стоит использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится  после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.

gennady-bars commented on src/components/app/App.jsx:
```diff
@@ -9,51 +16,47 @@ import OrderDetails from '../order-details/OrderDetails';
+          dispatch(orderInfo(data))
+        })
+        .catch(error => {
+          console.error('Error:', error)
```
Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, 

* Приложение ломается при отправке заказа на сервер https://disk.yandex.ru/i/X9-bwmX9R-1q_A 

Еще заметил ошибки:


* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты. 
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера
* Не стоит использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится  после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.
* Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`

yakovlevyaroslav committed:
Внес правки после первой проверки:
- вывел повторяющиеся куски кода запросов в фукции checkResponse и requestFunc
- добавил дополнительную проверку на наличие поля order в state order (может из-за ее отсутствия выдает ошибку при отображении модалки заказа)
- вывел урл апи в переменную BASE_URL
- переписал пост запрос postBurgerData c логикой thunk и перенес его в order/actions

gennady-bars commented on src/components/burger-consctructor/BurgerConstructor.jsx:
```diff
@@ -1,73 +1,121 @@
+        <div className={`${burgerConstructorStyles['burger-constructor_block']} ${burgerConstructorStyles['burger-constructor_scroll']}`}>
+          {otherIngredients.length > 0 ? (
+            otherIngredients.map((data, index) => (
+              <ConstructorItem key={`${data._id}-${index}`} id={data._id} index={index} data={data} />
```
```js

 key={`${data._id}-${index}`} 
```

Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**

gennady-bars replied:
Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

А если используете `createSlice`, то это делается в `prepare`

```ts

prepare: (ingredient: TIngredient) => {
        return { payload: { ...ingredient, uniqueId: uuid4() } };
      }


```

Подробнее про `prepare` тут https://redux-toolkit.js.org/api/createSlice#customizing-generated-action-creators

gennady-bars replied:
[Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)

gennady-bars commented on src/services/ingredients/actions.js:
```diff
@@ -0,0 +1,34 @@
+    .then(data => {
+      dispatch(fetchIngredientsSuccess(data.data));
+    })
+    .catch(error => dispatch(fetchIngredientsFailure(error.message)))
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Хорошая логика редакса


но есть некоторые недочеты:



* Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**


**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа

yakovlevyaroslav committed:
Поправил замечания из 2го комментария:
- добавил в код библиотеку uuid, для генерации уникального id в компоненте ConstructorItem
- Добавил очистку списка ингридиентов для бургера в заказе при успешной отправке запроса
- Добавил надпись Идет загрузк..., пока получаем номер заказа с бэкенда
- Добавил actions CLEAR_INGREDIENT_LIST, START_LOADING, STOP_LOADING для задач выше

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

yakovlevyaroslav commented:
> Поздравляю! Ваша работа принята.
> 
> Вы отлично потрудились.
> 
> Удачного дальнейшего обучения.

Спасибо за ревьи и за пожелание)
